src/bindings.d.ts

```diff
@@ -1,7 +1,47 @@
-import type { R2Bucket } from '@cloudflare/workers-types'
+import type { Link } from 'multiformats/link'
+import type { Context } from '@web3-storage/gateway-lib'
+import type { CARLink } from 'cardex/api'
+import type { R2Bucket, KVNamespace, RateLimit } from '@cloudflare/workers-types'
+import type { MemoryBudget } from './lib/mem-budget'
+import { CID } from '@web3-storage/gateway-lib/handlers'
+
+export {}
 
 export interface Environment {
+  VERSION: string
   DEBUG: string
   CARPARK: R2Bucket
   CONTENT_CLAIMS_SERVICE_URL?: string
+  RATE_LIMITS_SERVICE_URL?: string
+  ACCOUNTING_SERVICE_URL: string
+  RATE_LIMITER: RateLimit
+  AUTH_TOKEN_METADATA: KVNamespace
+  FF_RATE_LIMITER_ENABLED: boolean
+}
+
+export type GetCIDRequestData = Pick<Request, 'url' | 'headers'>
+
+export type GetCIDRequestOptions = GetCIDRequestData
+
+export interface RateLimitsService {
+  check: (cid: CID, options: GetCIDRequestOptions) => Promise<RateLimitExceeded>
+}
+
+export interface TokenMetadata {
+  locationClaim?: unknown // TODO: figure out the right type to use for this - we probably need it for the private data case to verify auth
+  invalid?: boolean
 }
+
+export interface RateLimits {
+  create: ({ env }: { env: Environment }) => RateLimitsService
```

Comment:
I think we don't generally define types for like this as we'd never pass this around in code. Instead this would just be a file with an exported function called create.

Reply:
I think it will be better if we use separate files for each middleware/handler instead of defining everything in `middleware.js`, otherwise it will be hard to maintain all of them in one place. Then I can refactor this as well.